src/mbgl/tile/geometry_tile_worker.cpp

```diff
@@ -143,24 +154,46 @@ void GeometryTileWorker::setPlacementConfig(PlacementConfig placementConfig_, ui
     }
 }
 
+void GeometryTileWorker::onGlyphsAvailable(GlyphPositionMap glyphs) {
+    glyphPositions = std::move(glyphs);
+    waitingForGlyphs = false;
+    symbolDependenciesChanged();
+}
+
+void GeometryTileWorker::onIconsAvailable(IconAtlasMap icons_) {
+    icons = std::move(icons_);
+    waitingForIcons = false;
+    symbolDependenciesChanged();
+}
+
+bool GeometryTileWorker::hasPendingSymbolDependencies() const {
+    return waitingForGlyphs || waitingForIcons;
+}
+
 void GeometryTileWorker::symbolDependenciesChanged() {
     try {
         switch (state) {
-        case Idle:
-            if (hasPendingSymbolDependencies()) {
+        case NeedPlacement:
+            if (!hasPendingSymbolDependencies()) {
```

Comment:
`setPlacementConfig` needs this guard as well for `case Idle`. Or alternatively (my preference): remove the places `coalesce` is currently called in favor of inlining it at the end of `attemptPlacement`. `attemptPlacement` already has an early exit for the `hasPendingSymbolDependencies` case, so it can then be called unconditionally.

Along with this, I also suggest factoring the followup to `redoLayout` that's currently repeated in four places:

```c++
            if (!hasPendingSymbolDependencies()) {
                coalesce();
            } else {
                state = NeedPlacement;
            }
```

Move that into `redoLayout` while relying on the fact that `attemptPlacement` now itself calls `coalesce`:

```c++
void GeometryTileWorker::redoLayout() {
    ...

    if (hasPendingSymbolDependencies()) {
        state = NeedPlacement;
    } else {
        attemptPlacement();
    }
}
```